File: apis/inferences/qwen3/sampling_manager.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-

# Copyright (c) 2025 HOUMO AI
#
# File: sampling_manager.py
# Description:
#   Sampling Manager for Qwen3 - Implements various sampling strategies including
#   top-k, top-p, temperature scaling, and repetition penalty for text generation.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0

import numpy as np
from typing import List, Optional


class SamplingManager:
    """
    Manages different sampling strategies for text generation.

    This class provides methods for applying various sampling techniques like
    temperature scaling, top-k, top-p (nucleus sampling), and repetition penalty
    to control the randomness and quality of generated text.
    """

    def __init__(
        self,
        temperature: float = 1.0,
        top_k: Optional[int] = None,
        top_p: float = 1.0,
        repetition_penalty: float = 1.0,
        min_tokens_to_keep: int = 1,
    ):
        """
        Initialize the sampling manager with specified parameters.

        Args:
            temperature: Controls randomness of predictions (higher = more random)
            top_k: Keep only k most likely tokens (None means no limit)
            top_p: Keep tokens with cumulative probability up to p (nucleus sampling)
            repetition_penalty: Penalty for repeated tokens (1.0 = no penalty)
            min_tokens_to_keep: Minimum number of tokens to keep during top-p filtering
        """
        self.temperature = temperature
        self.top_k = top_k
        self.top_p = top_p
        self.repetition_penalty = repetition_penalty
        self.min_tokens_to_keep = min_tokens_to_keep

    def softmax(self, x: np.ndarray) -> np.ndarray:
        """
        Compute softmax values for array x.

        Args:
            x: Input array of logits

        Returns:
            Softmax-normalized probabilities
        """
        exp_x = np.exp(x - np.max(x))
        return exp_x / np.sum(exp_x)

    def apply_temperature(self, logits: np.ndarray) -> np.ndarray:
        """
        Apply temperature scaling to logits.

        Args:
            logits: Input logits array

        Returns:
            Temperature-adjusted logits
        """
        if self.temperature <= 0:
            raise ValueError("Temperature must larger than 0")

        return logits / self.temperature

    def apply_repetition_penalty(
        self, logits: np.ndarray, previous_tokens: Optional[List[int]] = None
    ) -> np.ndarray:
        """
        Apply repetition penalty to logits based on previously generated tokens.

        Args:
            logits: Input logits array
            previous_tokens: List of previously generated token IDs

        Returns:
            Repetition penalty-adjusted logits
        """
        if self.repetition_penalty == 1.0 or not previous_tokens:
            return logits

        adjusted_logits = logits.copy()
        # Apply penalty to each unique token in the history
        for token_id in set(previous_tokens):
            if 0 <= token_id < len(logits):
                if logits[token_id] < 0:
                    adjusted_logits[token_id] = (
                        logits[token_id] * self.repetition_penalty
                    )
                else:
                    adjusted_logits[token_id] = (
                        logits[token_id] / self.repetition_penalty
                    )

        return adjusted_logits

    def apply_top_k(self, probs: np.ndarray) -> np.ndarray:
        """
        Apply top-k sampling to probability distribution.

        Args:
            probs: Probability distribution

        Returns:
            Top-k filtered probability distribution
        """
        if self.top_k is None or self.top_k <= 0:
            return probs

        top_k = min(self.top_k, len(probs))
        if top_k <= 0:
            return probs

        # Find indices of top-k probabilities
        top_k_indices = np.argpartition(probs, -top_k)[-top_k:]
        # Create mask to zero out non-top-k probabilities
        mask = np.ones_like(probs, dtype=bool)
        mask[top_k_indices] = False
        filtered_probs = probs.copy()
        filtered_probs[mask] = 0

        # Renormalize the probabilities
        if np.sum(filtered_probs) > 0:
            normalized_probs = filtered_probs / np.sum(filtered_probs)
        else:
            normalized_probs = np.ones_like(probs) / len(probs)

        return normalized_probs

    def apply_top_p(self, probs: np.ndarray) -> np.ndarray:
        """
        Apply top-p (nucleus) sampling to probability distribution.

        Args:
            probs: Probability distribution

        Returns:
            Top-p filtered probability distribution
        """
        if self.top_p >= 1.0:
            return probs

        # Sort probabilities in descending order
        sorted_indices = np.argsort(probs)[::-1]
        sorted_probs = probs[sorted_indices]

        cumulative_probs = np.cumsum(sorted_probs)
        cutoff_indices = np.where(cumulative_probs >= self.top_p)[0]

        if len(cutoff_indices) > 0:
            cutoff_index = cutoff_indices[0]
            if cutoff_index < self.min_tokens_to_keep - 1:
                cutoff_index = self.min_tokens_to_keep - 1

            selected_indices = sorted_indices[: cutoff_index + 1]
        else:
            selected_indices = sorted_indices

        # Zero out probabilities outside the selected set
        mask = np.ones_like(probs, dtype=bool)
        mask[selected_indices] = False
        filtered_probs = probs.copy()
        filtered_probs[mask] = 0

        # Renormalize the probabilities
        if np.sum(filtered_probs) > 0:
            normalized_probs = filtered_probs / np.sum(filtered_probs)
        else:
            normalized_probs = np.ones_like(probs) / len(probs)

        return normalized_probs

    def process_logits(
        self, logits: np.ndarray, previous_tokens: Optional[List[int]] = None
    ) -> np.ndarray:
        """
        Process raw logits through all configured sampling techniques.

        Args:
            logits: Raw logits from the model
            previous_tokens: List of previously generated token IDs

        Returns:
            Processed probability distribution ready for sampling
        """
        processed_logits = logits.copy()
        # Apply repetition penalty
        processed_logits = self.apply_repetition_penalty(
            processed_logits, previous_tokens
        )

        # Apply softmax
        # Note: Using direct processing instead of softmax for performance
        probs = processed_logits
        # probs = self.softmax(processed_logits)

        # Apply top-k filtering
        probs = self.apply_top_k(probs)

        # Apply top-p filtering
        probs = self.apply_top_p(probs)

        # Apply temperature scaling
        probs = self.apply_temperature(probs)

        return probs

    def sample(
        self, logits: np.ndarray, previous_tokens: Optional[List[int]] = None
    ) -> int:
        """
        Sample a token from the processed logits.

        Args:
            logits: Raw logits from the model
            previous_tokens: List of previously generated token IDs

        Returns:
            Sampled token ID as a numpy array
        """
        logits = logits[0]
        probs = self.process_logits(logits, previous_tokens)
        # Handle edge case where all probabilities are zero
        if np.all(probs == 0):
            probs = np.ones_like(probs) / len(probs)

        # sampled_index = np.random.choice(len(probs), p=probs)
        sampled_index = probs.argmax(-1)

        return np.array([[sampled_index]])

    def get_processed_probs(
        self, logits: np.ndarray, previous_tokens: Optional[List[int]] = None
    ) -> np.ndarray:
        """
        Get the processed probability distribution without sampling.

        Args:
            logits: Raw logits from the model
            previous_tokens: List of previously generated token IDs

        Returns:
            Processed probability distribution
        """
        return self.process_logits(logits, previous_tokens)

File: models/asr/whisper/cpp/src/SamplingManager.cc
/*
 * Copyright (c) 2025 HOUMO AI
 *
 * File: SamplingManager.cc
 * Description:
 *   Implementation of SamplingManager for text generation post-processing.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

#include "SamplingManager.h"

#include <algorithm>
#include <cmath>
#include <limits>
#include <map>
#include <numeric>
#include <utility>
#include <vector>

void SamplingManager::softmax(const float* logits, size_t size,
                              float* probs) const {
  if (size == 0) return;

  // Find max for numerical stability
  float max_val = logits[0];
  for (size_t i = 1; i < size; ++i) {
    if (logits[i] > max_val) max_val = logits[i];
  }

  // Compute exp(x - max) and sum
  float sum = 0.0f;
  for (size_t i = 0; i < size; ++i) {
    probs[i] = std::exp(logits[i] - max_val);
    sum += probs[i];
  }

  // Normalize
  if (sum > 0.0f) {
    for (size_t i = 0; i < size; ++i) {
      probs[i] /= sum;
    }
  }
}

void SamplingManager::applyTemperature(float* logits, size_t size) const {
  if (temperature_ <= 0.0f) {
    // Invalid temperature, skip
    return;
  }

  for (size_t i = 0; i < size; ++i) {
    logits[i] /= temperature_;
  }
}

void SamplingManager::applyRepetitionPenalty(
    float* logits, size_t size, const std::vector<int>& previous_tokens) const {
  if (repetition_penalty_ == 1.0f || previous_tokens.empty()) {
    return;
  }

  // Create unique set of previous tokens
  std::vector<int> unique_tokens = previous_tokens;
  std::sort(unique_tokens.begin(), unique_tokens.end());
  unique_tokens.erase(std::unique(unique_tokens.begin(), unique_tokens.end()),
                      unique_tokens.end());

  for (int token_id : unique_tokens) {
    if (token_id >= 0 && static_cast<size_t>(token_id) < size) {
      if (logits[token_id] < 0) {
        // Negative logits: multiply by penalty
        logits[token_id] *= repetition_penalty_;
      } else {
        // Positive logits: divide by penalty
        logits[token_id] /= repetition_penalty_;
      }
    }
  }
}

void SamplingManager::applyPresenceRepetitionPenalty(
    float* logits, size_t size, const std::vector<int>& previous_tokens) const {
  if (repetition_penalty_ == 0.0f || previous_tokens.empty()) {
    return;
  }

  // Create unique set of previous tokens
  std::vector<int> unique_tokens = previous_tokens;
  std::sort(unique_tokens.begin(), unique_tokens.end());
  unique_tokens.erase(std::unique(unique_tokens.begin(), unique_tokens.end()),
                      unique_tokens.end());

  // Apply presence penalty: subtract penalty from logits of repeated tokens
  for (int token_id : unique_tokens) {
    if (token_id >= 0 && static_cast<size_t>(token_id) < size) {
      logits[token_id] -= repetition_penalty_;
    }
  }
}

void SamplingManager::applyTopK(float* probs, size_t size) const {
  if (top_k_ <= 0 || static_cast<size_t>(top_k_) >= size) {
    return;  // Disabled or top_k >= vocab_size
  }

  size_t k = std::min(static_cast<size_t>(top_k_), size);

  // Find the k-th largest value using partial sort
  std::vector<float> temp_probs(probs, probs + size);
  std::nth_element(temp_probs.begin(), temp_probs.begin() + (size - k),
                   temp_probs.end(), std::greater<float>());
  float threshold = temp_probs[size - k];

  // Filter: set all values below threshold to 0
  float sum = 0.0f;
  for (size_t i = 0; i < size; ++i) {
    if (probs[i] < threshold) {
      probs[i] = 0.0f;
    } else {
      sum += probs[i];
    }
  }

  // Renormalize
  if (sum > 0.0f) {
    for (size_t i = 0; i < size; ++i) {
      if (probs[i] > 0.0f) {
        probs[i] /= sum;
      }
    }
  } else {
    // Fallback: uniform distribution
    for (size_t i = 0; i < size; ++i) {
      probs[i] = 1.0f / size;
    }
  }
}

void SamplingManager::applyTopP(float* probs, size_t size) const {
  if (top_p_ >= 1.0f) {
    return;  // Disabled
  }

  // Create index-probability pairs for sorting
  std::vector<std::pair<float, size_t>> indexed_probs(size);
  for (size_t i = 0; i < size; ++i) {
    indexed_probs[i] = {probs[i], i};
  }

  // Sort by probability descending
  std::sort(
      indexed_probs.begin(), indexed_probs.end(),
      [](const std::pair<float, size_t>& a, const std::pair<float, size_t>& b) {
        return a.first > b.first;
      });

  // Find cutoff index where cumulative probability >= top_p
  float cumulative = 0.0f;
  size_t cutoff_index = 0;
  for (size_t i = 0; i < size; ++i) {
    cumulative += indexed_probs[i].first;
    cutoff_index = i;
    if (cumulative >= top_p_) {
      break;
    }
  }

  // Ensure minimum tokens to keep
  size_t min_keep = static_cast<size_t>(min_tokens_to_keep_);
  if (cutoff_index < min_keep - 1) {
    cutoff_index = min_keep - 1;
  }

  // Build mask for tokens to keep
  std::vector<bool> keep_mask(size, false);
  for (size_t i = 0; i <= cutoff_index; ++i) {
    keep_mask[indexed_probs[i].second] = true;
  }

  // Filter: set values outside top-p to 0
  float sum = 0.0f;
  for (size_t i = 0; i < size; ++i) {
    if (!keep_mask[i]) {
      probs[i] = 0.0f;
    } else {
      sum += probs[i];
    }
  }

  // Renormalize
  if (sum > 0.0f) {
    for (size_t i = 0; i < size; ++i) {
      if (probs[i] > 0.0f) {
        probs[i] /= sum;
      }
    }
  } else {
    // Fallback: uniform distribution
    for (size_t i = 0; i < size; ++i) {
      probs[i] = 1.0f / size;
    }
  }
}

void SamplingManager::processLogits(const float* logits, size_t size,
                                    const std::vector<int>& previous_tokens,
                                    float* processed_probs) const {
  // Copy logits to working array
  std::vector<float> working_logits(logits, logits + size);

  // 1. Apply presence repetition penalty (matching Python implementation)
  applyPresenceRepetitionPenalty(working_logits.data(), size, previous_tokens);

  // 2. Use logits directly as probs (matching Python: not using softmax)
  // Python comment: "not using softmax in case of long time cost"
  std::copy(working_logits.begin(), working_logits.end(), processed_probs);

  // 3. Apply top-K
  applyTopK(processed_probs, size);

  // 4. Apply top-P
  applyTopP(processed_probs, size);

  // 5. Apply temperature
  applyTemperature(processed_probs, size);
}

int SamplingManager::sample(const float* logits, size_t size,
                            const std::vector<int>& previous_tokens) const {
  std::vector<float> probs(size);
  processLogits(logits, size, previous_tokens, probs.data());

  // Handle all-zero case
  bool all_zero = true;
  for (size_t i = 0; i < size; ++i) {
    if (probs[i] != 0.0f) {
      all_zero = false;
      break;
    }
  }
  if (all_zero) {
    for (size_t i = 0; i < size; ++i) {
      probs[i] = 1.0f / size;
    }
  }

  // Greedy sampling: return argmax (matching Python implementation)
  int sampled_index = 0;
  float max_prob = probs[0];
  for (size_t i = 1; i < size; ++i) {
    if (probs[i] > max_prob) {
      max_prob = probs[i];
      sampled_index = static_cast<int>(i);
    }
  }

  return sampled_index;
}

void SamplingManager::getProcessedProbs(const float* logits, size_t size,
                                        const std::vector<int>& previous_tokens,
                                        float* probs) const {
  processLogits(logits, size, previous_tokens, probs);
}